crates/ra_hir/src/ty/infer.rs

```diff
@@ -1045,18 +1045,32 @@ impl<'a, D: HirDatabase> InferenceContext<'a, D> {
             Expr::Lambda { body, args, arg_types } => {
                 assert_eq!(args.len(), arg_types.len());
 
+                let mut sig_tys = Vec::new();
+
                 for (arg_pat, arg_type) in args.iter().zip(arg_types.iter()) {
                     let expected = if let Some(type_ref) = arg_type {
                         self.make_ty(type_ref)
                     } else {
                         Ty::Unknown
                     };
-                    self.infer_pat(*arg_pat, &expected, BindingMode::default());
+                    let arg_ty = self.infer_pat(*arg_pat, &expected, BindingMode::default());
+                    sig_tys.push(arg_ty);
                 }
 
-                // FIXME: infer lambda type etc.
-                let _body_ty = self.infer_expr(*body, &Expectation::none());
-                Ty::Unknown
+                // add return type
+                let ret_ty = self.new_type_var();
+                sig_tys.push(ret_ty.clone());
+                let sig_ty = Ty::apply(
+                    TypeCtor::FnPtr { num_args: sig_tys.len() as u16 - 1 },
+                    sig_tys.into(),
+                );
+                let closure_ty = Ty::apply_one(
+                    TypeCtor::Closure { def: self.body.owner(), expr: tgt_expr },
+                    sig_ty,
+                );
+
+                self.infer_expr(*body, &Expectation::has_type(ret_ty));
+                closure_ty
```

Comment:
Note that closures need special handing when they syntactically appear as arguments: https://github.com/intellij-rust/intellij-rust/blob/be0a9d92cfaf06135b406463b65e409f77b35f8c/src/main/kotlin/org/rust/lang/core/types/infer/TypeInferenceWalker.kt#L689-L715

Reply:
Hmm... I'll need a failing test case before I implement that though :sweat_smile:

Reply:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=0c8d5323846d84a0ed2e6e72a0d0714f

Reply:
Oh man... that's the kind of thing I'd be happier off not knowing :smile: